common/cache/simple.go

```diff
@@ -0,0 +1,212 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cache
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	// SimpleCacheFull is the error that is returned from PutIfNotExists if cache is full
+	SimpleCacheFull = errors.New("simple cache is full")
+
+	// DummyCreateTime is the create time used by all entries in the cache.
+	DummyCreateTime = time.Now()
+)
+
+type (
+	simple struct {
+		sync.RWMutex
+		accessMap   map[interface{}]*list.Element
+		iterateList *list.List
+		maxSize     int
+		rmFunc      RemovedFunc
+	}
+
+	simpleItr struct {
+		simple   *simple
+		nextItem *list.Element
+	}
+
+	simpleEntry struct {
+		key   interface{}
+		value interface{}
+	}
+)
+
+// Close closes the iterator
+func (it *simpleItr) Close() {
+	it.simple.RUnlock()
+}
+
+// HasNext return true if there is more items to be returned
+func (it *simpleItr) HasNext() bool {
+	return it.nextItem != nil
+}
+
+// Next returns the next item
+func (it *simpleItr) Next() Entry {
+	if it.nextItem == nil {
+		panic("Simple cache iterator Next called when there is no next item")
+	}
+
+	entry := it.nextItem.Value.(*simpleEntry)
+	it.nextItem = it.nextItem.Next()
+	// make a copy of the entry so there will be no concurrent access to this entry
+	entry = &simpleEntry{
+		key:   entry.key,
+		value: entry.value,
+	}
+	return entry
+}
+
+func (e *simpleEntry) Key() interface{} {
+	return e.key
+}
+
+func (e *simpleEntry) Value() interface{} {
+	return e.value
+}
+
+// CreateTime is not implemented for simple cache entries
+func (e *simpleEntry) CreateTime() time.Time {
+	return DummyCreateTime
+}
+
+// NewSimple creates a new simple cache with given options.
+// Simple cache will never evict entries and it will never reorder the elements.
```

Comment:
So if it hits the upper limit of the cache, does it mean all other domains(not load into the cache) will be blocked?

Reply:
If cache fills up we will just start panicing.

Reply:
Yes. Panic makes sense to me.

Reply:
Why not just make this unbounded? This is for the expected unbounded use case so let it grow and hit potential oom is better than just panicing imo, as usually caller has no control on # of items to put to the cache in this case.

Reply:
Also, worth adding a gauge metrics on actual cache size. Need to be careful on where to emit the metrics though.

Reply:
+1 for removing the size limit.

Reply:
okay changed this.